test/common/shared_pool/shared_pool_test.cc

```diff
@@ -0,0 +1,168 @@
+#include <thread>
+
+#include "common/event/timer_impl.h"
+#include "common/shared_pool/shared_pool.h"
+
+#include "test/mocks/event/mocks.h"
+#include "test/test_common/thread_factory_for_test.h"
+#include "test/test_common/utility.h"
+
+#include "absl/synchronization/notification.h"
+#include "gtest/gtest.h"
+
+namespace Envoy {
+namespace SharedPool {
+
+class SharedPoolTest : public testing::Test {
+protected:
+  SharedPoolTest() : api_(Api::createApiForTest()), dispatcher_(api_->allocateDispatcher()) {
+    dispatcher_thread_ = api_->threadFactory().createThread([this]() {
+      // Must create a keepalive timer to keep the dispatcher from exiting.
+      std::chrono::milliseconds time_interval(500);
+      keepalive_timer_ = dispatcher_->createTimer(
+          [this, time_interval]() { keepalive_timer_->enableTimer(time_interval); });
+      keepalive_timer_->enableTimer(time_interval);
+      dispatcher_->run(Event::Dispatcher::RunType::Block);
+    });
+  }
+
+  ~SharedPoolTest() override {
+    dispatcher_->exit();
+    dispatcher_thread_->join();
+  }
+
+  void createObjectSharedPool(std::shared_ptr<ObjectSharedPool<int>>& pool) {
+    absl::Notification go;
+    dispatcher_->post([&pool, &go, this]() {
+      pool = std::make_shared<ObjectSharedPool<int>>(*dispatcher_);
+      go.Notify();
+    });
+    go.WaitForNotification();
+  }
+
+  void getObjectFromObjectSharedPool(std::shared_ptr<ObjectSharedPool<int>>& pool,
+                                     std::shared_ptr<int>& o, int value) {
+    absl::Notification go;
+    dispatcher_->post([&pool, &o, &go, value]() {
+      o = pool->getObject(value);
+      go.Notify();
+    });
+    go.WaitForNotification();
+  }
+
+  Api::ApiPtr api_;
+  Event::DispatcherPtr dispatcher_;
+  Thread::ThreadPtr dispatcher_thread_;
+  Event::TimerPtr keepalive_timer_;
+  absl::Notification go_;
+};
+
+TEST_F(SharedPoolTest, Basic) {
+  Event::MockDispatcher dispatcher;
```

Comment:
Can we just use the real dispatcher/pool for all tests in this file? It will make it much more likely to catch issues with TSAN, etc.